File: src/main/java/com/ject/studytrip/image/domain/util/ImageUrlUtil.java
package com.ject.studytrip.image.domain.util;

import static org.springframework.util.StringUtils.hasText;

import java.net.URI;
import java.util.Optional;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public final class ImageUrlUtil {
    // 이미지 최종 경로 생성
    public static String build(String baseUrl, String key) {
        if (!hasText(baseUrl) || !hasText(key)) {
            return null;
        }

        String normalizedDomain =
                baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;

        return normalizedDomain + "/" + key.trim();
    }

    // 이미지 최종 경로에서 이미지 키 추출
    public static Optional<String> extractKey(String baseUrl, String url) {
        if (url == null || url.isBlank()) {
            return Optional.empty();
        }

        if (baseUrl == null || baseUrl.isBlank()) {
            return Optional.empty();
        }

        // URI 파싱
        URI uri;
        try {
            uri = URI.create(url);
        } catch (IllegalArgumentException e) {
            return Optional.empty();
        }

        String host = uri.getHost();
        if (host == null || !host.equals(extractHostName(baseUrl))) {
            return Optional.empty();
        }

        String path = uri.getPath();
        if (path == null || path.length() <= 1) {
            return Optional.empty();
        }

        // 맨 앞의 '/' 제거
        String key = path.startsWith("/") ? path.substring(1) : path;
        return key.isBlank() ? Optional.empty() : Optional.of(key);
    }

    // 경로에서 Host 추출
    private static String extractHostName(String baseUrl) {
        if (baseUrl.startsWith("https://")) {
            return baseUrl.substring(8).replaceAll("/$", "");
        }
        if (baseUrl.startsWith("http://")) {
            return baseUrl.substring(7).replaceAll("/$", "");
        }
        return baseUrl.replaceAll("/$", "");
    }
}

File: src/main/kotlin/com/ject/studytrip/image/domain/util/ImageUrlUtil.kt
package com.ject.studytrip.image.domain.util

import java.net.URI

object ImageUrlUtil {
    // 이미지 최종 경로 생성
    fun build(
        baseUrl: String,
        key: String,
    ): String {
        require(baseUrl.isNotBlank())
        require(key.isNotBlank())

        val normalizedDomain = baseUrl.removeSuffix("/")
        return "$normalizedDomain/${key.trim()}"
    }

    // 이미지 URL에서 이미지 key 추출
    fun extractKey(
        baseUrl: String?,
        url: String?,
    ): String? {
        if (baseUrl.isNullOrBlank() || url.isNullOrBlank()) {
            return null
        }

        val uri =
            try {
                URI.create(url)
            } catch (_: IllegalArgumentException) {
                return null
            }

        val baseHost = extractHostName(baseUrl) ?: return null
        if (uri.host != baseHost) {
            return null
        }

        val path = uri.path ?: return null
        if (path.length <= 1) {
            return null
        }

        return path
            .removePrefix("/")
            .takeIf { it.isNotBlank() }
    }

    // baseUrl에서 host 추출
    private fun extractHostName(baseUrl: String): String? =
        try {
            URI.create(baseUrl).host
        } catch (_: IllegalArgumentException) {
            null
        }
}
